### ISSUE TITLE
Unresolved imports with hybrid implicit namespace packages

### ISSUE THREAD

**Issue by vfazio** [microsoft/pyright#5297]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
Starting in 1.1.308 and above, hybrid namespace packages start to fail in import resolution.
PEP 420 says this mixed mode should be supported, and it worked previously, so it seems like a regression.
This is despite a solution having been found according to `verbose` output:

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ pyright --verbose test.py
WARNING: there is a new pyright version available (v1.1.308 -> v1.1.314).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

Search paths for /mnt/development/test-project
  /home/vfazio/.cache/pyright-python/1.1.308/node_modules/pyright/dist/typeshed-fallback/stdlib
  /mnt/development/test-project
  /mnt/development/test-project/typings
  /home/vfazio/.cache/pyright-python/1.1.308/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /home/vfazio/.pyenv/versions/3.10.6/lib/python3.10
  /home/vfazio/.pyenv/versions/3.10.6/lib/python3.10/lib-dynload
  /mnt/development/test-project/venv/lib/python3.10/site-packages
  /mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg
Found 1 source file
Could not import 'ns.modulea' in file '/mnt/development/test-project/test.py'
  Looking in stubPath '/mnt/development/test-project/typings'
  Attempting to resolve stub package using root path '/mnt/development/test-project/typings'
  Attempting to resolve using root path '/mnt/development/test-project/typings'
  Looking in root directory of execution environment '/mnt/development/test-project'
  Attempting to resolve stub package using root path '/mnt/development/test-project'
  Attempting to resolve using root path '/mnt/development/test-project'
  Resolved import with file '/mnt/development/test-project/ns/modulea/__init__.py'
  Finding python search paths
  Executing interpreter: 'python3'
  Skipping '/home/vfazio/.pyenv/versions/3.10.6/lib/python311.zip' because it is not a valid directory
  Received 4 paths from interpreter
    /home/vfazio/.pyenv/versions/3.10.6/lib/python3.10
    /home/vfazio/.pyenv/versions/3.10.6/lib/python3.10/lib-dynload
    /mnt/development/test-project/venv/lib/python3.10/site-packages
    /mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg
  Looking in python search path '/home/vfazio/.pyenv/versions/3.10.6/lib/python3.10'
  Attempting to resolve stub package using root path '/home/vfazio/.pyenv/versions/3.10.6/lib/python3.10'
  Attempting to resolve using root path '/home/vfazio/.pyenv/versions/3.10.6/lib/python3.10'
  Looking in python search path '/home/vfazio/.pyenv/versions/3.10.6/lib/python3.10/lib-dynload'
  Attempting to resolve stub package using root path '/home/vfazio/.pyenv/versions/3.10.6/lib/python3.10/lib-dynload'
  Attempting to resolve using root path '/home/vfazio/.pyenv/versions/3.10.6/lib/python3.10/lib-dynload'
  Looking in python search path '/mnt/development/test-project/venv/lib/python3.10/site-packages'
  Attempting to resolve stub package using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages'
  Attempting to resolve using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages'
  Resolved import with file '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns/__init__.py'
  Looking in python search path '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg'
  Attempting to resolve stub package using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg'
  Attempting to resolve using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg'
  Resolved import with file '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg/ns/modulea/__init__.py'
  Looking for typeshed stdlib path
  Looking for typeshed stdlib path
  Attempting to resolve using root path '/home/vfazio/.cache/pyright-python/1.1.308/node_modules/pyright/dist/typeshed-fallback/stdlib'
  Typeshed path not found
  Looking for typeshed third-party path
  Looking for typeshed stubs path
  Typeshed path not found
pyright 1.1.308
/mnt/development/test-project/test.py
  /mnt/development/test-project/test.py:1:8 - error: Import "ns.modulea" could not be resolved (reportMissingImports)
1 error, 0 warnings, 0 informations 
Completed in 0.39sec
```

Note the `Resolved import with file '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg/ns/modulea/__init__.py'`

**To Reproduce**
The steps to reproduce are a bit complex, I'll do my best to be exhaustive.

Create a package that uses the legacy namespace syntax:

```
vfazio@vfazio4 ~/development/test-project2 $ tree ns
ns
├── __init__.py
└── moduleb
    └── __init__.py

1 directory, 2 files
```
```
vfazio@vfazio4 ~/development/test-project2 $ cat ns/__init__.py 
from pkgutil import extend_path
__path__ = extend_path(__path__, __name__)

```
with a simple setup file:

```
vfazio@vfazio4 ~/development/test-project2 $ cat setup.py 
"""Python setup.py for project_name package"""
from setuptools import find_namespace_packages, setup

setup(
    name="ns_modb",
    version="0.1",
    description="project_description",
    url="https://github.com/author_name/project_urlname/",
    long_description="no description",
    author="author_name",
    packages=find_namespace_packages(exclude=["tests", ".github"]),
    entry_points={
        "console_scripts": ["project_name = project_name.__main__:main"]
    },

```

Build the wheel `python3 setup.py bdist_wheel`

Create a second package (namespace A) that uses the "newer" implicit namespace syntax (without an __init__.py):

```
vfazio@vfazio4 ~/development/test-project $ tree ns
ns
└── modulea
    └── __init__.py
```

Similar setup.py:

```
"""Python setup.py for project_name package"""
from setuptools import find_namespace_packages, setup

setup(
    name="ns_moda",
    version="0.1",
    description="project_description",
    url="https://github.com/author_name/project_urlname/",
    long_description="no description",
    author="author_name",
    packages=find_namespace_packages(exclude=["tests", ".github"]),
    entry_points={
        "console_scripts": ["project_name = project_name.__main__:main"]
    },
)
```

Create a virtual env and install pyright and namespace module B

```
vfazio@vfazio4 /mnt/development/test-project $ python3 -m venv venv
vfazio@vfazio4 /mnt/development/test-project $ . venv/bin/activate

(venv) vfazio@vfazio4 /mnt/development/test-project $ python3 -c "import sys; print(sys.path)"
['', '/usr/lib/python310.zip', '/usr/lib/python3.10', '/usr/lib/python3.10/lib-dynload', '/mnt/development/test-project/venv/lib/python3.10/site-packages']


(venv) vfazio@vfazio4 /mnt/development/test-project $ pip install pyright==1.1.307
```

Install namespace module A by either:
```
python3 setup.py install
```
or 
```
echo `pwd` > venv/lib/python3.10/site-packages/pkg.pth
```

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ pip list
Package    Version
---------- -------
nodeenv    1.8.0
ns-moda    0.1
ns-modb    0.1
pip        22.3.1
pyright    1.1.307
setuptools 65.5.0

```

Insure the package egg or path shows up in `sys.path`

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ python3 -c "import sys; print(sys.path)"
['', '/usr/lib/python310.zip', '/usr/lib/python3.10', '/usr/lib/python3.10/lib-dynload', '/mnt/development/test-project/venv/lib/python3.10/site-packages', '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg']
```

or

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ python3 -c "import sys; print(sys.path)"
['', '/usr/lib/python310.zip', '/usr/lib/python3.10', '/usr/lib/python3.10/lib-dynload', '/mnt/development/test-project/venv/lib/python3.10/site-packages', '/mnt/development/test-project']
```

Create a small file to test imports

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ cat test.py 
import ns.modulea

from ns import moduleb
```

With pyright 1.1.307:

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ pyright --verbose test.py 
WARNING: there is a new pyright version available (v1.1.307 -> v1.1.314).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

Search paths for /mnt/development/test-project
  /home/vfazio/.cache/pyright-python/1.1.307/node_modules/pyright/dist/typeshed-fallback/stdlib
  /mnt/development/test-project
  /mnt/development/test-project/typings
  /home/vfazio/.cache/pyright-python/1.1.307/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/lib/python3.10
  /usr/lib/python3.10/lib-dynload
  /mnt/development/test-project/venv/lib/python3.10/site-packages
  /mnt/development/test-project
  /mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg
Found 1 source file
pyright 1.1.307
0 errors, 0 warnings, 0 informations 
Completed in 0.398sec
```

With 1.1.308:

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ pyright --verbose test.py 
WARNING: there is a new pyright version available (v1.1.308 -> v1.1.314).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

Search paths for /mnt/development/test-project
  /home/vfazio/.cache/pyright-python/1.1.308/node_modules/pyright/dist/typeshed-fallback/stdlib
  /mnt/development/test-project
  /mnt/development/test-project/typings
  /home/vfazio/.cache/pyright-python/1.1.308/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/lib/python3.10
  /usr/lib/python3.10/lib-dynload
  /mnt/development/test-project/venv/lib/python3.10/site-packages
  /mnt/development/test-project
  /mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg
Found 1 source file
Could not import 'ns.modulea' in file '/mnt/development/test-project/test.py'
  Looking in stubPath '/mnt/development/test-project/typings'
  Attempting to resolve stub package using root path '/mnt/development/test-project/typings'
  Attempting to resolve using root path '/mnt/development/test-project/typings'
  Looking in root directory of execution environment '/mnt/development/test-project'
  Attempting to resolve stub package using root path '/mnt/development/test-project'
  Attempting to resolve using root path '/mnt/development/test-project'
  Resolved import with file '/mnt/development/test-project/ns/modulea/__init__.py'
  Finding python search paths
  Executing interpreter: 'python3'
  Skipping '/usr/lib/python310.zip' because it is not a valid directory
  Received 5 paths from interpreter
    /usr/lib/python3.10
    /usr/lib/python3.10/lib-dynload
    /mnt/development/test-project/venv/lib/python3.10/site-packages
    /mnt/development/test-project
    /mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg
  Looking in python search path '/usr/lib/python3.10'
  Attempting to resolve stub package using root path '/usr/lib/python3.10'
  Attempting to resolve using root path '/usr/lib/python3.10'
  Looking in python search path '/usr/lib/python3.10/lib-dynload'
  Attempting to resolve stub package using root path '/usr/lib/python3.10/lib-dynload'
  Attempting to resolve using root path '/usr/lib/python3.10/lib-dynload'
  Looking in python search path '/mnt/development/test-project/venv/lib/python3.10/site-packages'
  Attempting to resolve stub package using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages'
  Attempting to resolve using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages'
  Resolved import with file '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns/__init__.py'
  Looking in python search path '/mnt/development/test-project'
  Attempting to resolve stub package using root path '/mnt/development/test-project'
  Attempting to resolve using root path '/mnt/development/test-project'
  Resolved import with file '/mnt/development/test-project/ns/modulea/__init__.py'
  Looking in python search path '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg'
  Attempting to resolve stub package using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg'
  Attempting to resolve using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg'
  Resolved import with file '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg/ns/modulea/__init__.py'
  Looking for typeshed stdlib path
  Looking for typeshed stdlib path
  Attempting to resolve using root path '/home/vfazio/.cache/pyright-python/1.1.308/node_modules/pyright/dist/typeshed-fallback/stdlib'
  Typeshed path not found
  Looking for typeshed third-party path
  Looking for typeshed stubs path
  Typeshed path not found
pyright 1.1.308
/mnt/development/test-project/test.py
  /mnt/development/test-project/test.py:1:8 - error: Import "ns.modulea" could not be resolved (reportMissingImports)
1 error, 0 warnings, 0 informations 
Completed in 0.391sec
```

with 1.1.314

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ pyright --verbose test.py 
Search paths for /mnt/development/test-project
  /home/vfazio/.cache/pyright-python/1.1.314/node_modules/pyright/dist/typeshed-fallback/stdlib
  /mnt/development/test-project
  /mnt/development/test-project/typings
  /home/vfazio/.cache/pyright-python/1.1.314/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/lib/python3.10
  /usr/lib/python3.10/lib-dynload
  /mnt/development/test-project/venv/lib/python3.10/site-packages
  /mnt/development/test-project
  /mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg
Found 1 source file
Could not import 'ns.modulea' in file '/mnt/development/test-project/test.py'
  Looking in stubPath '/mnt/development/test-project/typings'
  Attempting to resolve stub package using root path '/mnt/development/test-project/typings'
  Attempting to resolve using root path '/mnt/development/test-project/typings'
  Looking in root directory of execution environment '/mnt/development/test-project'
  Attempting to resolve stub package using root path '/mnt/development/test-project'
  Attempting to resolve using root path '/mnt/development/test-project'
  Resolved import with file '/mnt/development/test-project/ns/modulea/__init__.py'
  Finding python search paths
  Executing interpreter: 'python3'
  Skipping '/usr/lib/python310.zip' because it is not a valid directory
  Received 5 paths from interpreter
    /usr/lib/python3.10
    /usr/lib/python3.10/lib-dynload
    /mnt/development/test-project/venv/lib/python3.10/site-packages
    /mnt/development/test-project
    /mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg
  Looking in python search path '/usr/lib/python3.10'
  Attempting to resolve stub package using root path '/usr/lib/python3.10'
  Attempting to resolve using root path '/usr/lib/python3.10'
  Looking in python search path '/usr/lib/python3.10/lib-dynload'
  Attempting to resolve stub package using root path '/usr/lib/python3.10/lib-dynload'
  Attempting to resolve using root path '/usr/lib/python3.10/lib-dynload'
  Looking in python search path '/mnt/development/test-project/venv/lib/python3.10/site-packages'
  Attempting to resolve stub package using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages'
  Attempting to resolve using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages'
  Resolved import with file '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns/__init__.py'
  Looking in python search path '/mnt/development/test-project'
  Attempting to resolve stub package using root path '/mnt/development/test-project'
  Attempting to resolve using root path '/mnt/development/test-project'
  Resolved import with file '/mnt/development/test-project/ns/modulea/__init__.py'
  Looking in python search path '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg'
  Attempting to resolve stub package using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg'
  Attempting to resolve using root path '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg'
  Resolved import with file '/mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg/ns/modulea/__init__.py'
  Looking for typeshed stdlib path
  Looking for typeshed stdlib path
  Attempting to resolve using root path '/home/vfazio/.cache/pyright-python/1.1.314/node_modules/pyright/dist/typeshed-fallback/stdlib'
  Typeshed path not found
  Looking for typeshed third-party path
  Looking for typeshed stubs path
  Typeshed path not found
pyright 1.1.314
/mnt/development/test-project/test.py
  /mnt/development/test-project/test.py:1:8 - error: Import "ns.modulea" could not be resolved (reportMissingImports)
1 error, 0 warnings, 0 informations 
Completed in 0.422sec
```

As a sanity check, nuke the hybrid `__init__.py` installed by moduleb from the site-package and see that it works.

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ cat venv/lib/python3.10/site-packages/ns/__init__.py 
from pkgutil import extend_path
__path__ = extend_path(__path__, __name__)

(venv) vfazio@vfazio4 /mnt/development/test-project $ rm venv/lib/python3.10/site-packages/ns/__init__.py 
(venv) vfazio@vfazio4 /mnt/development/test-project $ pyright --verbose test.py 
Search paths for /mnt/development/test-project
  /home/vfazio/.cache/pyright-python/1.1.314/node_modules/pyright/dist/typeshed-fallback/stdlib
  /mnt/development/test-project
  /mnt/development/test-project/typings
  /home/vfazio/.cache/pyright-python/1.1.314/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/lib/python3.10
  /usr/lib/python3.10/lib-dynload
  /mnt/development/test-project/venv/lib/python3.10/site-packages
  /mnt/development/test-project
  /mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg
Found 1 source file
pyright 1.1.314
0 errors, 0 warnings, 0 informations 
Completed in 0.428sec
```

**Expected behavior**

Behavior should work as it did in 1.1.307 and prior:

```
(venv) vfazio@vfazio4 /mnt/development/test-project $ pyright --verbose test.py
WARNING: there is a new pyright version available (v1.1.307 -> v1.1.313).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

Search paths for /mnt/development/test-project
  /home/vfazio/.cache/pyright-python/1.1.307/node_modules/pyright/dist/typeshed-fallback/stdlib
  /mnt/development/test-project
  /mnt/development/test-project/typings
  /home/vfazio/.cache/pyright-python/1.1.307/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /home/vfazio/.pyenv/versions/3.10.6/lib/python3.10
  /home/vfazio/.pyenv/versions/3.10.6/lib/python3.10/lib-dynload
  /mnt/development/test-project/venv/lib/python3.10/site-packages
  /mnt/development/test-project/venv/lib/python3.10/site-packages/ns_moda-0.1-py3.10.egg
Found 1 source file
pyright 1.1.307
0 errors, 0 warnings, 0 informations 
Completed in 0.397sec
```

**VS Code extension or command-line**
This does affect Pylance as it bundles pyright. Currently I'm having to explicitly install 2023.5.30 which has pyright 1.1.307. Newer versions of pylance that bundle pyright are broken for our use case.

**Additional context**
Just to preempt the "update the other python libraries to not include __init__.py" chorus; that's unfortunately not something we can do easily and definitely not something we can do quickly given the large number of historical python packages we have.

**Comment by erictraut** [comment]
Pyright does work with cases that involve both traditional (static) and implicit namespace packages. It will not work with packages that dynamically modify the `__path__` variable, as you're doing above.

If it happened to work in previous versions of pyright, then you were just getting lucky. I fixed a bug in 1.1.308 that would explain the reason that you saw a change in behavior.

I recommend that you eliminate the use of dynamic `__path__` manipulation in your code. This technique is ancient and hasn't been necessary since the introduction of PEP 420.

If you comment out the dynamic `__path__` modification code and run the code, you should find that pyright's new (post 1.1.308) behavior accurately reflects the runtime behavior. We have no plans to try to accommodate dynamic `__path__` modifications.

**Comment by vfazio** [comment]
That's an unfortunate response. As mentioned at the end of the issue, our company has a lot of legacy code that "just works", so now, instead of accommodating hybrid packages, which is allowed for in the PEP, in order to utilize pyright, and thus, pylance, we need to update all of our packages, or switch language servers.
